### ISSUE TITLE
Iteration step_size should accept an iterable

### ISSUE THREAD

**Issue by YSelfTool** [scikit-hep/uproot4#174]
For processing a single tree in a single ROOT file with multiple processes, it's necessary for those processes to read disjoint parts of the tree. In uproot3, this was possible by passing an iterator as `entrysteps` to `iterate`.

Currently, this should be possible by using `HasBranches.iterate` directly (setting ob the objects manually) and passing `entry_start` and `entry_stop`.

The simplest solution might be to add another parameter `entry_skip` or something like that, passed through the iterate functions, and replace 
```
for sub_entry_start in uproot4._util.range(
    entry_start, entry_stop, entry_step
):
```
by
```
for sub_entry_start in uproot4._util.range(
    entry_start, entry_stop, entry_step * entry_skip
):
```
in `HasBranches.iterate`.

Alternatively, step_size could become and iterator that defaults to that range, or a function directly replacing range here.

**Comment by jpivarski** [comment]
I'll let `step_size` be an iterable of `(start, stop)` pairs, like it was in Uproot 3. That's the simplest solution.

**Comment by jpivarski** [comment]
I've changed my mind about this one. I was just going to implement it because I was recently looking at the implementation to fix a memory leak (PR #187). But what I saw was that there's no advantage to passing a list of `(start, stop)` pairs through `iterate` over just doing

```python
for start, stop in list_of_pairs:
    arrays = ttree.arrays(entry_start=start, entry_stop=stop, array_cache=None)
```

The performance advantages of using `ttree.iterate` (and `uproot4.iterate`) rely on it being sequential, since it saves `TBaskets` from the previous iteration to avoid rereading them on the next. (PR #187 fixes a broken implementation of that.) But if the `list_of_pairs` can skip entries, there's a (good) chance that the previously saved `TBaskets` won't be relevant, and if the `list_of_pairs` can skip around non-monotonically (not your case, I know), then the previously saved `TBaskets` definitely aren't relevant.

I could add an interface that just does the above—if (badly named) `step_size` is an iterable, then just call `arrays` a bunch of times—but that interface wouldn't generalize to `uproot4.iterate`. Implementing it on `HasBranches.iterate` but not `uproot4.iterate` complicates the interface.

So I'd rather recommend that you write the for loop calling `HasBranches.arrays` a bunch of times. I think that is the most future-friendly way to go.
